lib/utils/batcher/state.go

```diff
@@ -69,11 +80,20 @@ type StateMonitor struct {
 
 	mu    sync.RWMutex
 	state State
+	// Timer used to auto-exit Overloaded after a quiet period.
+	flushTimer clockwork.Timer
```

Comment:
There's no need to "live update" the state with a goroutine fired off by a timer, store the time at which the StateMonitor became overloaded and reset it on read instead.

Reply:
If nobody reads the state why do we need to update it? I am very confused.

Reply:
Where login running gorutine that does the GetState from the client side will take a memory where the transition from Normal->Overloaded -> normal state are quire rare so my first pick was time.After call instead of long living goroutine.

Reply:
I don't think an asynchronous state change - because a `time.AfterFunc` goroutine _is_ asynchronous - is what we should be using. If there's a need for this sort of control we need the loop that ingests events to make the decision to start ignoring events and wait for calmer times, it should not be some asynchronous timer in a state machine (that until then was purely deterministic and based on data) that trips and triggers a callback out again to the reader.

Reply:
> ignoring events and wait for calmer times

But how do you know if there is calmer times without receiving any event ? 

You need either monitor the last timestamp in gorutine or have wakeup timer that checks that. 

This is exacly what this logic does. The AfterFunc alternative is having a asychronios gorutine that monitor the state so basically it the same to the current approach.

Reply:
The fact that `OnExitOverloaded` also needs to handle being called from a different goroutine at the same time as it's getting called through `UpdateState` also makes things much harder than they ought to be.

Reply:
>  being driven by hidden timers

This is explicit flag in the Config 

> The fact that OnExitOverloaded also needs to handle being called from a different goroutine at the same time as it's getting called through UpdateState also makes things much harder than they ought to be.

I think that it simplify things. 

The caller is not needed to manage previous state next state transition and can specify the action on state transition.